File: lib/domain/db/models/TimelinePin.dart
import 'package:cobble/domain/db/converters/SqlJsonConverters.dart';
import 'package:cobble/domain/db/models/NextSyncAction.dart';
import 'package:cobble/domain/db/models/TimelinePinType.dart';
import 'package:json_annotation/json_annotation.dart';
import 'package:uuid_type/uuid_type.dart';

import 'TimelinePinLayout.dart';

part 'TimelinePin.g.dart';

@JsonSerializable()
@UuidConverter()
@BooleanNumberConverter()
@NumberDateTimeConverter()
class TimelinePin {
  /// Unique UUID of the item
  final Uuid itemId;

  /// UUID of the watchapp that owns this pin
  final Uuid parentId;

  /// ID of the item that is backing this pin
  /// (for example ID of the calendar event)
  /// can be null
  final String backingId;

  /// Date of the pin start
  final DateTime timestamp;

  /// Duration of the event in minutes
  final int duration;

  /// Type of the pin
  final TimelinePinType type;

  /// ???
  final bool isVisible;

  /// ???
  final bool isFloating;

  /// Whether pin spans throughout the whole day (Duration ignored???)
  final bool isAllDay;

  /// ???
  final bool persistQuickView;

  /// UI Layout of the pin
  final TimelinePinLayout layout;

  /// JSON of the timeline pin attributes. Those attributes must
  /// correspond to the specified [layout].
  final String attributesJson;

  /// JSON of the pin actions. Can be *null* if pin has no actions.
  final String actionsJson;

  /// Action that should be performed for this pin
  /// when the next sync-to-watch is performed
  final NextSyncAction nextSyncAction;

  TimelinePin(
      this.itemId,
      this.parentId,
      this.backingId,
      this.timestamp,
      this.duration,
      this.type,
      this.isVisible,
      this.isFloating,
      this.isAllDay,
      this.persistQuickView,
      this.layout,
      this.attributesJson,
      this.actionsJson,
      this.nextSyncAction);

  Map<String, dynamic> toMap() {
    return _$TimelinePinToJson(this);
  }

  factory TimelinePin.fromMap(Map<String, dynamic> map) {
    return _$TimelinePinFromJson(map);
  }

  @override
  String toString() {
    return 'TimelinePin{itemId: $itemId, parentId: $parentId, backingId: $backingId, timestamp: $timestamp, duration: $duration, type: $type, isVisible: $isVisible, isFloating: $isFloating, isAllDay: $isAllDay, persistQuickView: $persistQuickView, layout: $layout, attributesJson: $attributesJson, actionsJson: $actionsJson, nextSyncAction: $nextSyncAction}';
  }
}

File: android/shared/src/commonMain/kotlin/io/rebble/cobble/shared/database/entity/TimelinePin.kt
package io.rebble.cobble.shared.database.entity

import androidx.room.Entity
import androidx.room.PrimaryKey
import com.benasher44.uuid.Uuid
import io.rebble.cobble.shared.database.NextSyncAction
import io.rebble.libpebblecommon.packets.blobdb.TimelineItem
import kotlinx.datetime.Clock
import kotlinx.datetime.Instant
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.json.Json
import kotlin.time.Duration.Companion.days
import kotlin.time.Duration.Companion.minutes

@Entity
data class TimelinePin(
        /**
         * UUID of the item
         */
        @PrimaryKey val itemId: Uuid,
        /**
         * UUID of the watchapp that owns this item
         */
        val parentId: Uuid,
        /**
         * ID of the item that backs this item e.g. a calendar event ID
         */
        val backingId: String?,
        /**
         * Date of pin start
         */
        val timestamp: Instant,
        /**
         * Duration in minutes
         */
        val duration: Int?,
        val type: TimelineItem.Type,
        /**
         * Doesn't seem to be used
         */
        val isVisible: Boolean = true,
        /**
         * Display pin in UTC timezone
         */
        val isFloating: Boolean = false,
        /**
         * Ignores duration, just displays the pin for the whole day
         */
        val isAllDay: Boolean = false,
        /**
         * Shows the event in quick view while active
         */
        val persistQuickView: Boolean = false,
        val layout: TimelineItem.Layout,
        /**
         * JSON of timeline pin attributes, corresponding to the [layout]
         */
        val attributesJson: String?,
        /**
         * JSON of timeline pin actions, null if no actions
         */
        val actionsJson: String?,
        val nextSyncAction: NextSyncAction?
)

val TimelinePin.isInPast: Boolean
    get() = if (isAllDay) {
        timestamp + 1.days < Clock.System.now()
    } else {
        timestamp + (duration?.minutes ?: 0.minutes) < Clock.System.now()
    }